aten/src/ATen/cuda/CUDAGraph.cpp

```diff
@@ -11,6 +11,27 @@
 
 namespace at::cuda {
 
+namespace {
+cudaStream_t create_external_stream() {
+  // From:
+  // https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1g793d7d4e474388ddfda531603dc34aa3
+  // "Capture must be ended on the same stream in which it was initiated, and it
+  // may only be initiated if the stream is not already in capture mode."
+
+  // Since pytorch uses a pool of 32 pre-allocated cuda streams,
+  // should a user nest 32 conditional nodes, there would be an error
+  // for the 32nd node, since that node's stream would already be in
+  // capture mode. The easiest solution is to handle stream creation
+  // and deletion ourselves.
+
+  // Be sure to call cudaStreamDestroy on this when it is finished
```

Comment:
It is inside `end_capture_to_conditional_node`. https://github.com/pytorch/pytorch/pull/140979/files#diff-d7302d133bb5e0890fc94de9aeea4d9d442555a3b40772c9db10edb5cf36a35cR561

Another possibility is to make this return a unique_ptr<cudaStream_t> with a custom destructor. That would be a little bit awkward since the `std::stack<at::cuda::CUDAStreamGuard> conditional_node_streams_` would need an additional stack just containing these unique pointers (CUDAStreamGuard lacks a move constructor, so I'm not sure that having a std::pair of CUDAStreamGuard and a unique_ptr would work here).

Reply:
So I did try your suggestion: https://github.com/pytorch/pytorch/pull/140979/commits/7cdced5139de61172aca2ba681cc67c54f0a7b39

To be honest, I think it makes things more opaque and confusing in this instance to use smart pointers. In particular, I almost accidentally depended upon the ordering of calculation of the arguments of a function here (std::move(child_stream) means that *child_stream might return NULL): https://github.com/pytorch/pytorch/pull/140979/commits/7cdced5139de61172aca2ba681cc67c54f0a7b39#diff-d7302d133bb5e0890fc94de9aeea4d9d442555a3b40772c9db10edb5cf36a35cR456-R462

I'm thinking of reverting the change personally, but would be happy to keep it if you feel strongly about it.